doc generator: improve title, add meta description

inoas opened the pull request:
This helps:

- discoverabilty of gleam libs via search engines
- ranking on search engines before there is less indications for duplicate content
- users organizing many tabs and toggling through them (`<title>`)
- accessibility for screen readers that use the title

Example, stdlib 0.27, homepage:

```html
<title>gleam_stdlib · v0.27.0</title>
<meta name="description" content="gleam_stdlib v0.27.0 · A standard library for the Gleam programming language"/>
```

Example, stdlib 0.27, string module page:

```html
<title>gleam/string · gleam_stdlib · 0.27.0</title>
<meta name="description" content="gleam_stdlib v0.27.0 · gleam/string · Strings in Gleam are UTF-8 binaries. They can be written in your code as text surrounded by &amp;quot;double quotes&amp;quot;."/>
```

inoas commented:
I have also added support for non-readme.md pages, such as license or changelog which can be linked.

## Example

```toml
[documentation]
pages = [
	{ title = "CHANGELOG", path = "changelog.html", source = "CHANGELOG.md" },
	{ title = "LICENSE", path = "license.html", source = "LICENSE" },
]
```

```html
<title>glychee · v0.2.7 · license</title>
<meta name="description" content="glychee v0.2.7 · Glychee: Easy access to Elixir&#x27;s Benchee from Gleam!"/>
```

```html
<title>glychee · v0.2.7 · changelog</title>
<meta name="description" content="glychee v0.2.7 · Glychee: Easy access to Elixir&#x27;s Benchee from Gleam!"/>
```

lpil commented on compiler-core/src/docs.rs:
```diff
@@ -206,6 +214,19 @@ pub fn generate_html(
+            .as_str(),
+            "\n",
+            " ",
+        );
```
I like the idea of having a description, but the module doc comment could be thousands of lines long- I don't think it is suitable for this tag. What do other languages use here?

inoas replied:
We don't really have anything else. We can truncate after 320 (or 500) characters with one elipsis character?

> … aka U+2026

inoas replied:
> view-source:https://hexdocs.pm/phoenix/Phoenix.Naming.html

lacks stuff. the best we have is the module doc introduction.

inoas replied:
... I have pushed an update, the code may be not idiomatic and I couldn't make `format!()` work for now... but I'd rather have your opinion first if the direction is good for you.

lpil replied:
We need to look at other options here, the doc comment is not suitable for use as a description.

inoas commented:
view-source:https://www.typescriptlang.org/docs/handbook/2/everyday-types.html
has `<meta data-react-helmet="true" name="description" content="The language primitives."/>`

inoas commented:
view-source:https://www.typescriptlang.org/docs/handbook/intro.html
has `<meta data-react-helmet="true" name="description" content="Your first step to learn TypeScript"/>`
... seems manual.

inoas commented on compiler-core/src/docs.rs:
```diff
@@ -206,6 +239,28 @@ pub fn generate_html(
+        let page_meta_description_truncated = if page_meta_description_len <= 320 {
+            &page_meta_description[0..page_meta_description_len]
+        } else {
+            // &format!("{}…", &page_meta_description[0..320 - 1])
```
Any idea how to do this?

inoas commented:
so the linter complains about my otherworldy bad rust skills... 
before tackling that... is this generally something you want?

inoas commented:
I have refactored the code and removed infos duplciate in meta description that should only appear in the title.

I have modified the output examples and included one for stdlib bool module on the opening post of this issue: https://github.com/gleam-lang/gleam/pull/2043#issue-1609733339

inoas commented on compiler-core/src/docs.rs:
```diff
@@ -77,20 +81,52 @@ pub fn generate_html(
+        };
+        let page_meta_description = match page_path_without_ext {
+            "index" => render_as_meta_description(config.description.to_string().clone()),
+            _other => "".to_owned(),
```
`_other => "".to_owned()` =>
Could also read markdown metadata in a future PR:

- https://stackoverflow.com/questions/44215896/markdown-metadata-format
- https://stackoverflow.com/questions/42952149/how-to-add-metadata-in-github-flavoured-markdown

inoas replied:
Is there interest for this? Should I create an issue for this?

lpil replied:
We'd need to come up with some designs prior to making an issue to work on. Gathering some options would be good

inoas commented on compiler-core/src/docs.rs:
```diff
@@ -202,19 +239,23 @@ pub fn generate_html(
         });
 
+        let page_title = format!("{} · {} · v{}", name, config.name, config.version);
+        let page_meta_description = "";
```
Here it would best if we could add an optional short meta description in `.gleam` module files @lpil?

inoas replied:
Should I create an issue for this?

inoas replied:
We could also colocate optional meta data in the file system - not sure how noisy that would get.

inoas commented:
I have removed the automatic meta descriptions. The only meta description now injected is for the index/README.

The html-title is now much more precise and should improve finding gleam libraries and/or documentation via search engines nonetheless.

The html-title changes also help for the accessibility and developer experience when having multiple tabs open.

lpil commented on CHANGELOG.md:
```diff
@@ -14,6 +14,9 @@
   expression within them.
+- Generated docs now export more meta data to improve the developer experience,
+  accessibility and search engine discoverability.
+- Added support for unary negation of integers.
```
I think this is an extra entry here

lpil commented on compiler-core/src/docs.rs:
```diff
@@ -77,20 +81,52 @@ pub fn generate_html(
+            _other => format!("{} · {} · v{}", page.title, config.name, config.version),
+        };
+        let page_meta_description = match page_path_without_ext {
+            "index" => render_as_meta_description(config.description.to_string().clone()),
```
I don't think we need to do any HTML manipulation stuff for the description field, just putting it straight in is OK

inoas commented:
> We'd need to come up with some designs prior to making an issue to work on. Gathering some options would be good

What do you mean? description for modules? for non-index pages? keywords?
Because the title and meta description changes still left, should be a straight up improvement...

lpil reviewed (approved):
Wonderful! Thank you very much 💜

lpil commented:
I meant for the per-page metadata 👍
